Proxies sampling documentation (#14521)
DataDog/documentation: content/en/tracing/setup_overview/proxy_setup/_index.md

      - prefix: "cluster.datadog_agent."
```

## Sampling

The sampling rate for traces starting from Envoy can be set between `0.0` (0%) and `1.0` (100%).

Configure the sampling rate in order to control the volume of traces sent to Datadog by using the parameter `DD_TRACE_SAMPLING_RULES`.
By default, 100% of traces starting from Envoy are sampled.

To use the [Datadog Agent calculated sampling rates][2] (10 traces per second per Agent) and ignore the default sampling rule set to 100%, set the parameter `DD_TRACE_SAMPLING_RULES` to an empty array:

```
DD_TRACE_SAMPLING_RULES=[]
```

You can also define an explicit sampling rate between `0.0` (0%) and `1.0` (100%) by service. For instance, to set the sample rate to 10% for service `envoy-proxy`, set :

```
DD_TRACE_SAMPLING_RULES=[{"service": "envoy-proxy","sample_rate": 0.1}]
```

### How to apply the DD_TRACE_SAMPLING_RULES configuration option ?

> If envoy is executed by a shell script, set the environment variable right before executing envoy in the shell script

Author: each proxy type is already a tab, we cannot make sub tabs i think :/

> If envoy is executed by a shell script, set the environment variable right before executing envoy in the shell script

Reviewer: correct, we cannot

      - prefix: "cluster.datadog_agent."
```

## Sampling

> The sampling rate for traces starting from Envoy can be set between `0.0` (0%) and `1.0` (100%).
> 
> Configure the sampling rate in order to control the volume of traces sent to Datadog by using the parameter `DD_TRACE_SAMPLING_RULES`.
> By default, 100% of traces starting from Envoy are sampled.

Reviewer: ```suggestion
To control the volume of traces starting from Envoy that are sent to Datadog, specify a sampling rate by setting the parameter `DD_TRACE_SAMPLING_RULES` to a value between `0.0` (0%) and `1.0` (100%). If no value is specified, 100% of traces starting from Envoy are sent.
```

To use the [Datadog Agent calculated sampling rates][2] (10 traces per second per Agent) and ignore the default sampling rule set to 100%, set the parameter `DD_TRACE_SAMPLING_RULES` to an empty array:

```
DD_TRACE_SAMPLING_RULES=[]
```

> You can also define an explicit sampling rate between `0.0` (0%) and `1.0` (100%) by service. For instance, to set the sample rate to 10% for service `envoy-proxy`, set :

Reviewer: ```suggestion
You can also define an explicit sampling rate between `0.0` (0%) and `1.0` (100%) by service. For example, to set the sample rate to 10% for service `envoy-proxy`:
```

```
DD_TRACE_SAMPLING_RULES=[{"service": "envoy-proxy","sample_rate": 0.1}]
```

> ### How to apply the DD_TRACE_SAMPLING_RULES configuration option ?

Reviewer: ```suggestion
### Setting the `DD_TRACE_SAMPLING_RULES` configuration option
```

> 

Reviewer: ```suggestion

Use one of the following methods to configure your sampling rate, depending on how you run Envoy:

```

> #### In a Shell script
> 
> If envoy is executed by a shell script, set the environment variable right before executing envoy in the shell script

Reviewer: ```suggestion
- **By shell script**: Set the environment variable immediately before executing `envoy` in the script:
```

```
#!/bin/sh
export DD_TRACE_SAMPLING_RULES=[]
envoy -c envoy-config.yaml
```

> #### Docker compose setup
> 
> If envoy is executed as a service in a docker-compose setup, set the environment variable in the `environment` section of the service definition:

Reviewer: ```suggestion

- **In a Docker Compose setup**: Set the environment variable in the `environment` section of the service definition:
```

```
services:
  envoy:
    image: envoyproxy/envoy:v1.19-latest
    entrypoint: []
    command:
        - envoy
        - -c
        - /etc/envoy/envoy.yaml
    volumes:
        - './envoy.yaml:/etc/envoy/envoy.yaml:ro'
    environment:
        - DD_TRACE_SAMPLING_RULES=[]
```

> #### Kubernetes pod container
> 
> If envoy is a container inside of a Kubernetes pod, speficy the env variable in the `env` section of the relevant `containers` entry of the pod's spec:

Reviewer: ```suggestion
- **As a container inside a Kubernetes pod**: specify the environment variable in the `env` section of the corresponding `containers` entry of the pod's spec:
```

| v1.9 | v0.3.6 |

[1]: https://github.com/DataDog/dd-opentracing-cpp/tree/master/examples/envoy-tracing

> [2]: tracing/trace_ingestion/mechanisms#in-the-agent

Reviewer: ```suggestion
[2]: /tracing/trace_ingestion/mechanisms#in-the-agent
```

```
The above overrides the default `nginx-ingress-controller.ingress-nginx` service name.

### Sampling

> The sampling rate for traces starting from Nginx services can be set between `0.0` (0%) and `1.0` (100%).
> 
> Configure the sampling rate in order to control the volume of traces sent to Datadog by using the parameter `DD_TRACE_SAMPLING_RULES`.
> By default, 100% of traces starting from Nginx are sampled (the Kubernetes Nginx ingress controller uses the [v1.2.1][8] of the `dd-opentracing-cpp` library).

Reviewer: ```suggestion
To control the volume of traces starting from Nginx that are sent to Datadog, specify a sampling rate by setting the parameter `DD_TRACE_SAMPLING_RULES` to a value between `0.0` (0%) and `1.0` (100%). If no value is specified, 100% of traces starting from Nginx are sent. The Kubernetes Nginx ingress controller uses [v1.2.1][8] of the `dd-opentracing-cpp` library).
```

To use the [Datadog Agent calculated sampling rates][9] (10 traces per second per Agent) and ignore the default sampling rule set to 100%, set the parameter `DD_TRACE_SAMPLING_RULES` to an empty array:

```
DD_TRACE_SAMPLING_RULES=[]
```

> Append a `main-snippet` data entry to the controller's ConfigMap, as below:

Reviewer: ```suggestion
To configure `DD_TRACE_SAMPLING_RULES`, add a `main-snippet` entry in the `data` section of the controller's ConfigMap:
```

For [CronJobs][8], the `app` label should be added to the job template, as the generated name comes from the `Job` instead
of the higher-level `CronJob`.

### Sampling

> The sampling rate for traces starting from Istio services can be set between `0.0` (0%) and `1.0` (100%).
> 
> Configure the sampling rate in order to control the volume of traces sent to Datadog by using the parameter `DD_TRACE_SAMPLING_RULES`.
> By default, 100% of traces starting from Istio are sampled.

Reviewer: ```suggestion
To control the volume of traces starting from Istio that are sent to Datadog, specify a sampling rate by setting the parameter `DD_TRACE_SAMPLING_RULES` to a value between `0.0` (0%) and `1.0` (100%). If no value is specified, 100% of traces starting from Istio are sent. 
```

To use the [Datadog Agent calculated sampling rates][9] (10 traces per second per Agent) and ignore the default sampling rule set to 100%, set the parameter `DD_TRACE_SAMPLING_RULES` to an empty array:

```
DD_TRACE_SAMPLING_RULES=[]
```

> In each deployment whose namespace is labeled istio-injection=enabled, the environment variable can be set as part of the apm.datadoghq.com/env annotation of the deployment spec template.

Reviewer: ```suggestion
To configure `DD_TRACE_SAMPLING_RULES`, in each deployment whose namespace is labeled `istio-injection=enabled`, set the environment variable as part of the `apm.datadoghq.com/env` annotation of the deployment spec template:
```
